### PR TITLE
kubelet: Add an option for default seccomp profile

### PR DESCRIPTION
**Pull request by stlaz**
**What type of PR is this?**
Uncomment only one, leave it on its own line:
/kind api-change
/sig auth

**What this PR does / why we need it**:
This PR adds the `--default-seccomp-profile` option to kubelet. This options sets the default seccomp profile for any pods run in such a kubelet.

**Special notes for your reviewer**:
This PR is based on the discussion at https://github.com/kubernetes/enhancements/issues/135

**Does this PR introduce a user-facing change?**:

```release-note
A new experimental option was added to kubelet: --default-seccomp-profile. This option sets the default seccomp profile for any container to be run on such a kubelet. If not set, it defaults to `runtime/default`.
```

**Additional notes:**
There's still code in that might be a subject to change:
https://github.com/kubernetes/kubernetes/blob/fda41d14c42d9b8e88933359629a6e02b115f8ba/pkg/kubelet/dockershim/helpers_linux.go#L51-L55

Right now it seems that there's no other way to propagate the default seccomp policy here when nothing is set in the seccomp profile we get from a container.

On one hand, that is fine, because in real life this should always be set. On the other hand, we can't remove the empty string special case as that would break tests, unless we inject the seccomp profile in the tests, which in turn breaks code isolation, so I'd rather retrain from that.

cc @tallclair

### CODE REVIEW

File: cmd/kubelet/app/options/options.go [hunk 1]
```diff
@@ -385,7 +388,8 @@ func (f *KubeletFlags) AddFlags(mainfs *pflag.FlagSet) {
 	fs.StringVar(&f.VolumePluginDir, "volume-plugin-dir", f.VolumePluginDir, "The full path of the directory in which to search for additional third party volume plugins")
 	fs.StringVar(&f.LockFilePath, "lock-file", f.LockFilePath, "<Warning: Alpha feature> The path to file for kubelet to use as a lock file.")
 	fs.BoolVar(&f.ExitOnLockContention, "exit-on-lock-contention", f.ExitOnLockContention, "Whether kubelet should exit upon lock-file contention.")
-	fs.StringVar(&f.SeccompProfileRoot, "seccomp-profile-root", f.SeccompProfileRoot, "<Warning: Alpha feature> Directory path for seccomp profiles.")
+	fs.StringVar(&f.SeccompProfileRoot, "seccomp-profile-root", f.SeccompProfileRoot, "[Experimental] Directory path for seccomp profiles.")
```
Comment (tallclair) [inline cmd/kubelet/app/options/options.go]: I think we should leave this description alone for now - experimental seems like a demotion from alpha

File: cmd/kubelet/app/options/options.go [hunk 2]
```diff
@@ -385,7 +388,8 @@ func (f *KubeletFlags) AddFlags(mainfs *pflag.FlagSet) {
...
+	fs.StringVar(&f.DefaultSeccompProfile, "default-seccomp-profile", f.DefaultSeccompProfile, "[Experimental] The default seccomp profile")
```
Comment (tallclair) [inline cmd/kubelet/app/options/options.go]: If it's experimental, should it be behind a feature flag? I'm not sure this needs to be considered experimental...

File: cmd/kubelet/app/options/options.go [hunk 3]
```diff
@@ -162,6 +162,8 @@ type KubeletFlags struct {
 	ExitOnLockContention bool
 	// seccompProfileRoot is the directory path for seccomp profiles.
 	SeccompProfileRoot string
+	// seccompDefaultProfile is the default profile to be used
```
Comment (tallclair) [inline cmd/kubelet/app/options/options.go]: Is this the path to the profile, or the serialized profile json itself? Make it clear.

File: cmd/kubelet/app/options/options.go (hunk 2 above)
Comment (tallclair) [inline cmd/kubelet/app/options/options.go]: ditto - make it clear whether this is the path to the file holding the default profile, or the profile itself.

**Review by tallclair — COMMENTED** [review]
How does this interact with setting the profile to `docker/default` or `runtime/default`?

**Comment by stlaz** [comment]
Thank you for the review. I updated the descriptions as well as replaced `[Experimental]` with just `<Warning: alpha feature>`. Would you like me to add a feature gate for this? I guess it might make sense since this effectively turns seccomp functionality on by default :thinking: 

Also it made me realize there was a bug which would not change the `localhost/PATH` properly to `localhost//SeccompRoot/PATH` for default profile in `getSeccompProfileFromAnnotations()`, that's now fixed and a unit test was added.

> How does this interact with setting the profile to `docker/default` or `runtime/default`?

No changes were done in handling this, such values get passed on as they would have done previously.

**Comment by wangzhen127** [comment]
>> How does this interact with setting the profile to docker/default or runtime/default?

> No changes were done in handling this, such values get passed on as they would have done previously.

What if the values conflict? For example, if `docker/default` is set in the annotation and this flag is pointing to a specific profile different from the default one?

This way of changing kubelet behavior directly is adding a new way of setting seccomp profile, and making it more complicated. Why do we want to do it this way?

**Comment by stlaz** [comment]
> What if the values conflict? For example, if `docker/default` is set in the annotation and this flag is pointing to a specific profile different from the default one?

@wangzhen127 The fields take priority, so the order should really be *Pod's annotation* > *PSP default* > *default from kubelet*. This option should not rewrite a profile that was already set, but rather fill a missing one. I guess I was not entirely accurate saying no change was done to this process.

> This way of changing kubelet behavior directly is adding a new way of setting seccomp profile, and making it more complicated. Why do we want to do it this way?

We may want this in case PSP admission controller is turned off, but the user would still like to run in a more secure mode either set by their container runtime defaults, the kube secure default (future), or the one they set themselves. Also, and possibly more importantly, this option should serve as a switch to turn back the "old" behavior (= "unconfined" as default) while we're setting the default seccomp to "runtime/default" (or, later "kube/default").

**Comment by wangzhen127** [comment]
/cc @yujuhong 
Adding kubelet owner @yujuhong to give some opinions here.

Who is the target user of this flag? I suspect that it is platform provider, who turns on this flag across the fleet. If so, they can also just turn on PSP admission controller, right?

**Comment by stlaz** [comment]
Both your points are correct. There's but a little catch with the PSPs, and that is you'd have to specify the default profile for each of those, otherwise it would be just unconfined (or whatever the hardcoded default would be).

**Pull request closed without merging**